lib/rubocop/cop/metrics/line_length.rb

```diff
@@ -40,6 +42,45 @@ def max
           cop_config['Max']
         end
 
+        def allow_heredoc?
+          allowed_heredoc
+        end
+
+        def allowed_heredoc
+          cop_config['AllowHeredoc']
+        end
+
+        def extract_heredocs(processed_source)
+          heredocs = []
+          iter = processed_source.tokens.each
+          loop do
+            token = iter.next
```

Comment:
@rrosenblum I actually first tried a similar approach to what you're suggesting. Turns out that makes it really hard to parse. The advantage of the iterator is that you consume it, thus progressing in the token list. This is much easier with nested heredocs. See e.g. this snippet:

```
$ ruby-parse -E -e "
foo(<<ONE, <<TWO)
Hello #{<<NESTED.upcase}
World
NESTED
ONE
What's your name?
TWO
"
foo(<<ONE, <<TWO)
^~~ tIDENTIFIER "foo"                           expr_cmdarg  [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
   ^ tLPAREN2 "("                               expr_beg     [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
    ^~~~~ tSTRING_BEG "<<\""                    expr_end     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
^~~~~~ tSTRING_CONTENT "Hello "                 expr_end     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
      ^~ tSTRING_DBEG "\#{"                     expr_end     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
        ^~~~~~~~ tSTRING_BEG "<<\""             expr_end     [0 <= cond] [0 <= cmdarg]
World
^ tSTRING_CONTENT "World\n"                     expr_end     [0 <= cond] [0 <= cmdarg]
NESTED
^~~~~~ tSTRING_END "NESTED"                     expr_end     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
                ^ tDOT "."                      expr_dot     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
                 ^~~~~~ tIDENTIFIER "upcase"    expr_arg     [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
                       ^ tSTRING_DEND "}"       interp_string [0 <= cond] [0 <= cmdarg]
Hello #{<<NESTED.upcase}
                        ^ tSTRING_CONTENT "\n"  expr_end     [0 <= cond] [0 <= cmdarg]
World
^ tSTRING_CONTENT "World\n"                     expr_end     [0 <= cond] [0 <= cmdarg]
NESTED
^ tSTRING_CONTENT "NESTED\n"                    expr_end     [0 <= cond] [0 <= cmdarg]
ONE
^~~ tSTRING_END "ONE"                           expr_end     [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
         ^ tCOMMA ","                           expr_beg     [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
           ^~~~~ tSTRING_BEG "<<\""             expr_end     [0 <= cond] [0 <= cmdarg]
What's your name?
^ tSTRING_CONTENT "What's your name?\n"         expr_end     [0 <= cond] [0 <= cmdarg]
TWO
^~~ tSTRING_END "TWO"                           expr_end     [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
                ^ tRPAREN ")"                   expr_end     [0 <= cond] [0 <= cmdarg]
foo(<<ONE, <<TWO)
                 ^ tNL nil                      line_begin   [0 <= cond] [0 <= cmdarg]

^ false "$eof"                                  line_begin   [0 <= cond] [0 <= cmdarg]
(send nil :foo
  (dstr
    (str "Hello ")
    (begin
      (send
        (str "World\n") :upcase))
    (str "\n")
    (str "World\n")
    (str "NESTED\n"))
  (str "What's your name?\n"))
```

Let's cut down the noise here and just look at the tokens:

 order                        | token             | text                    | line
------------------------------|-------------------|-------------------------|------
1                             | `tIDENTIFIER`     | `"foo"`                 | 1
2                             | `tLPAREN2`        | `"("`                   | 1
3                             | `tSTRING_BEG`     | `"<<\""`                | 1
4                             | `tSTRING_CONTENT` | `"Hello "`              | 2
5                             | `tSTRING_DBEG`    | `"\#{"`                 | 2
6                             | `tSTRING_BEG`     | `"<<\""`                | 2
7                             | `tSTRING_CONTENT` | `"World\n"`             | 3
8                             | `tSTRING_END`     | `"NESTED"`              | 4
9                             | `tDOT`            | `"."`                   | 2
10                            | `tIDENTIFIER`     | `"upcase"`              | 2
11                            | `tSTRING_DEND`    | `"}"`                   | 2
12                            | `tSTRING_CONTENT` | `"\n"`                  | 2
13                            | `tSTRING_CONTENT` | `"World\n"`             | 3
14                            | `tSTRING_CONTENT` | `"NESTED\n"`            | 4
15                            | `tSTRING_END`     | `"ONE"`                 | 5
16                            | `tCOMMA`          | `","`                   | 1
17                            | `tSTRING_BEG`     | `"<<\""`                | 1
18                            | `tSTRING_CONTENT` | `"What's your name?\n"` | 6
19                            | `tSTRING_END`     | `"TWO"`                 | 7
20                            | `tRPAREN`         | `")"`                   | 1

So every time you encounter a `tSTRING_BEG`, the next token will be the first line of the heredoc body. You do that until you reach `tSTRING_END`, which contains the delimiter of the heredoc. The line before that is the end of the heredoc's body. Inside the heredoc you might encounter a further `tSTRING_BEG` (token 5), i.e. a nested heredoc. There you want to consume the tokens until that nested one reaches its end (token 8) before going back to the parent heredoc.

I feel like a solution using `#each` and the like would be more complicated since you'd have to pass state around, e.g. how many lines to skip due to a nested heredoc. With the iterator the state is implicit in the iterator. But I'm happy to be proven wrong :smile: